draft-ietf-quic-transport.md

```diff
@@ -2882,6 +3020,83 @@ by a client in protected packets, because it is certain that the server is able
 to decipher the packet.
 
 
+## ACK_ECN Frame {#frame-ack-ecn}
+
+A QUIC connection MUST keep counters for each ECN codepoint, recording the
+number of packets that were received with the corresponding ECN codepoint in
+the IP header. If the header is not readable from the application, the
+codepoint 00 (Not-ECT) MUST be assumed.
+
+ACK_ECN Frame MUST be used when when an endpoint is acknowledging a packet were
+the IP header ECN field was marked as ECT(0), ECT(1) or ECN-CE when received.
+ACK Frames ({{frame-ack}}) MAY be sent when all packets to be acknowledged
+had an IP header with the ECN field marked as Not-ECT.
+
+The ACK_ECN frame is used by the receiver to echo the value of these counters
+back to the sender of these packets. This allows the sender to utilize these
+counter values for congestion control. The ACK_ECN frame contains all the
+elements of the ACK frame ({{frame-ack}}) with the addition of an ECN block
+appended at the end.
+
+
+~~~
+ 0                   1                   2                   3
+ 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|                     Largest Acknowledged (i)                ...
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|                          ACK Delay (i)                      ...
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|                       ACK Block Count (i)                   ...
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|                          ACK Blocks (*)                     ...
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|                           ECN Block                         ...
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+~~~
+{: #ACN_ECN_FRAME_FORMAT title="ACK_ECN Frame Format"}
+
+
+### ECN Block {#ECN-Block}
+
+The ECN block is described below. The size (i) indicates variable-length
+encoding, explained in {{integer-encoding}}. The encoding length for each
+counter (1, 2, 4 or 8 bytes) should be selected such that all the
+significant bits are represented.
+
+~~~
+ 0                   1                   2                   3
+ 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|              Number of ECT(0) marked packets (i)            ...
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|              Number of ECT(1) marked packets (i)            ...
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|              Number of ECN-CE marked packets (i)            ...
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+~~~
+{: #ECN-BLOCK-FIG title="ECN Block"}
+
+
+### ECN counters
+
+The receiver side should implement three 64-bit counters that are copied to the
+ECN block when an ACK_ECN frame is generated:
```

Comment:
Rather than phrase this as "implement these counters" just define the fields in the frame.